### PR TITLE
Add "agoric" adapter

### PR DESCRIPTION
**Pull request by michaelfig**
This is working end-to-end with https://github.com/smartcontractkit/external-initiator/pull/73 and https://github.com/Agoric/dapp-oracle

### CODE REVIEW

File: agoric/adapter.js
Comment (thodges-gh) [inline agoric/adapter.js:75]:
```suggestion
  const data = JSON.parse(request.data.result)
  request.data = data
  const validator = new Validator(request, customParams)
```
This should get you the data sent from a previous task.

Comment (thodges-gh) [reply agoric/adapter.js:75]: Better yet if you put a JsonParse adapter _before_ the Agoric adapter, you don't have to do the parsing in this adapter. You'll only have to grab the values you want from `request.data.result`.

Comment (michaelfig) [reply agoric/adapter.js:75]: Thanks, I got the `request.data.result` correctly already and returned it to the chain.  I still don't know how to affect the display of `'Task Run Data'` for the Agoric task, which always seems to be `''`.  If I reset the `request.data` before calling the validator, won't that break the other data items I need, like `data.request_id` and `data.payment`?

Screen Shot 2020-10-26 at 9 49 11 PM

**Comment by RodrigoAD** [comment]
Also @michaelfig, if you could please add the tests necessary would be great. We cannot merge until we have them.

File: agoric/index.js
Comment (michaelfig) [reply agoric/index.js]: My problem was in forking the adapter template from something other than `example`.

File: agoric/src/httpSender.ts
Comment (RodrigoAD) [inline agoric/src/httpSender.ts:14]: You can make use of the `Requester.request` function instead of using axios directly. It just a layer on top of `axios` with some helpers.

File: agoric/package.json
Comment (RodrigoAD) [inline agoric/package.json:33]: Remove the `axios` dependency, as we'll use the `Requester.request`  function

File: agoric/src/httpSender.ts
Comment (michaelfig) [reply agoric/src/httpSender.ts]: Done!

**Review by thodges-gh — CHANGES_REQUESTED** [review]
I think we'll just want to add this to the release build process.

**Comment by michaelfig** [comment]
> I think we'll just want to add this to the release build process.

I added `"agoric"` to `.github/strategy/adapters.json` but now I'm getting a CI failure that I'm not sure what's going on.

**Comment by michaelfig** [comment]
> > I think we'll just want to add this to the release build process.
> 
> I added `"agoric"` to `.github/strategy/adapters.json` but now I'm getting a CI failure that I'm not sure what's going on.

```
@chainlink/external-adapter
[...]
1 failing

  1) Requester
       Requester.request
         returns an error from an endpoint:

      AssertionError: expected 'timeout of 100ms exceeded' to equal 'Request failed with status code 500'
      + expected - actual

      -timeout of 100ms exceeded
      +Request failed with status code 500
      
      at Context.<anonymous> (test/requester.test.js:38:16)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Review by krebernisak — CHANGES_REQUESTED** [review]

**Comment by michaelfig** [comment]
@krebernisak thanks for your constructive review.

I will likely be able to address your comments this week, but I have some other things I need to do first.

File: agoric/src/httpSender.ts
Comment (michaelfig) [reply agoric/src/httpSender.ts:13]: Yes.

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:104]: Okay, but for whatever reason when I use the adapter with just a `throw e`, the Chainlink node's run status page does not propagate errors correctly for failures.

It just gives a very generic error, with no details.

This took me a long time to figure out, so I'm hesitant to remove this code unless you know that it will work correctly (i.e. propagate errors of any kind to the node UI) without it.

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:25]: The purpose of this code is to defend the invariants of the `getRequiredFee` code I have below.  Changing it to use `bn.js` as you suggested will make this obsolete.

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:49]: It would be harder to plumb through because in the actual execute body I want to validate, but in reporting an error I do not.

Comment (michaelfig) [reply agoric/src/adapter.ts:49]: Not so much, I don't want to retry (the endpoint is always localhost).

**Review by krebernisak — CHANGES_REQUESTED** [review]
@michaelfig thanks for your work on this, and your patience 👍 

I left a few more questions, but we should be able to get this in very soon.

I see that you followed our work on this repo, as some types and the structure around these adapters evolved in the last couple of months. Thanks for that! We are still not done with all the changes, so we'll also probably slightly update this adapter in the future, after merging this in.

File: agoric/src/adapter.ts
Comment (justinkaseman) [reply agoric/src/adapter.ts:89]: Likely related to #204

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:30]: > The localhost service is always open, no authentication needed?

That's correct.

File: agoric/README.md
Comment (michaelfig) [reply agoric/README.md:13]: The inbound payment parameter is how much the client has offered.

The outbound payment parameter is how much the node is going to take, capped at a maximum to be what the client offered.  That's protected by our chain's offer safety.  The remainder is returned to the client as a refund.

I don't know how to do it yet, but it should be possible for nodes to decide not to take as much as they were offered.

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:14]: They can't take more than what was offered.

File: agoric/src/adapter.ts
Comment (michaelfig) [reply agoric/src/adapter.ts:75]: Ahh yes, optional chaining.  I'll do that.

File: agoric/src/index.ts
Comment (michaelfig) [reply agoric/src/index.ts:11]: Okay, good advice.

**Comment by michaelfig** [comment]
@krebernisak R4R again.  Sorry I took so long to get to it.

File: agoric/package.json
Comment (RodrigoAD) [inline agoric/package.json:40]: We've reduced the BigNumbers libraries to only two. Please use `ethers.BigNumber` or `Decimal`

File: agoric/src/httpSender.ts
Comment (RodrigoAD) [inline agoric/src/httpSender.ts:13]: I don't think `makeHTTPSender` is really necessary, it's just wrapping the oracle url. I recommend you to follow the [external adapter example](https://github.com/smartcontractkit/external-adapters-js/tree/develop/example/src). You will see that all this configuration can be set under `config`.

File: agoric/src/adapter.ts
Comment (RodrigoAD) [inline agoric/src/adapter.ts:92]: Try to follow the external adapter example

**Review by RodrigoAD — COMMENTED** [review]
Hi Michael, thanks for your patience. I've left you a couple of comments, but overall we should update the PR to follow the rest of the repo style. 
This could serve you: [Example Adapter](https://github.com/smartcontractkit/external-adapters-js/tree/develop/example/src)

**Comment by michaelfig** [comment]
> I've left you a couple of comments, but overall we should update the PR to follow the rest of the repo style.

I've done the style update now.  It passes local tests.  When CI passes it will be ready for review.

File: agoric/src/index.ts
Comment (RodrigoAD) [inline agoric/src/index.ts:8]: Be sure to be up to date with develop, and run `yarn setup`. `util.wrapExecute` shouldn't be necessary

File: agoric/src/index.ts
Comment (RodrigoAD) [inline agoric/src/index.ts:6]: You could include this into `makeConfig`. No worry about extending the `Config` type

Comment (michaelfig) [reply agoric/src/index.ts]: Thanks, I had missed this in my self-review.

File: agoric/package.json
Comment (RodrigoAD) [reply agoric/package.json:45]: Don't forget to remove the dependency

**Comment by michaelfig** [comment]
@RodrigoAD thanks for your comments, this is R4R again.

File: agoric/package.json
Comment (RodrigoAD) [inline agoric/package.json:45]: @michaelfig Don't forget to remove the dependency

Comment (michaelfig) [reply agoric/package.json:45]: Argh.  Yes!

Comment (michaelfig) [reply agoric/package.json]: @RodrigoAD I've fixed this, and CI passes.  Hopefully this looks good-to-go.  R4R!

**Pull request merged**
